README.rst
==============
Steemit Boiler
==============

Python Steemit bot for automatically following people, making timely posts, maximizing curation rewards, and posting market summaries.


Dependencies
------------

Tested in Python 3.6 in Linux and OSX.

Requires the following Python modules installed:

* appdirs
* ag.logging (aka logpy: https://git.alphagriffin.com/AlphaGriffin/logpy)
* matplotlib (for market mode graphs)
* steemit


Setup
-----

To build and install this application to the local system::

    make all && sudo make install

The above command installs the python module and ``boiler`` command to your system.

Alternatively, you can run the application in-place without installing::

    ./boiler

Without any arguments this command will display the usage help.


Account Credentials
-------------------

When you first enter follow mode (see next section) you will be asked for your account credentials, specifically your Account ID (Steemit username) and Private Posting Key. To get your Private Posting Key, go to your account permission page (https://steemit.com/@your_username/permissions), and click "Show Private Key" for the "Posting" key. This is the key you will enter into Boiler when prompted.

WARNING: Boiler stores these credentials unencrypted in your user home directory. A future release will encrypt these files or utilize the `steempy` client for credential management.


Automatically Follow Users
--------------------------

To automatically follow users, you must specify one or more tags where boiler will look for new posts. For example, to follow all users posting in "bitcoin" and "life"::

    ./boiler follow bitcoin life


Timely Posting
--------------

This bot can re-post a post at a time you specify. Simply craft a normal post with the following additional characteristics:

1. **Tag it to your username**. Your username must be the only tag.
2. **Add timely data on the last line**. The last line of your post must contain the date, time and tags that define when and where it will be re-posted. Here's an example::

    This post will be re-posted on Christmas day at 4:20 PM to the 'spam' and 'testing' tags.

    2017-12-25 16:20 spam testing

When running the bot in timely mode, it will first scan your post history to see if any timely posts are due (or overdue). It will then continuously scan for any new posts that you make::

    ./boiler timely


Curate Posts
------------

This bot can attempt to maximize curation rewards by voting on posts that are between 27 and 30 minutes old, up to 11 votes per day. The algorithm looks for posts in the tags you specify with successively higher pending payouts::

    ./boiler curate bitcoin life


Summarize the Market
--------------------

This bot can automatically post a cryptocurrency market summary for any given currency pair. Poloniex is used for market data, so the pair must exist there. If there is an RPC error while posting (usually because you've exceeded your STEEM bandwidth allocation) then it will keep trying every minute for up to an hour. To summarize a market, simply invoke market mode with a currency pair, post title, and tag(s)::

    ./boiler market btc-usdt "Summarizing the Bitcoin Market" bitcoin spam

The title you specify will be appended with a colon character followed by the latest action.

You can also set the historical all-time-high for a currency pair. This only needs to be done once... if ``market`` mode is run regularly then the all-time-high will stay updated::

    ./boiler market-ath btc-usdt 1234.56


Support
-------

Alpha Griffin is just a couple guys working on cool projects in our spare time. Join our Gitter chat (https://gitter.im/AlphaGriffin/Lobby) for help, suggestions, or just to see what we're up to.

If you find this software useful, please consider a small STEEM donation to @lannocc .



CHANGELOG
Copyright (C) 2017 Alpha Griffin
@%@~LICENSE~@%@

saw_081317_5 - Don't save new all-time-high if in testing mode.

saw_081317_4 - Market mode Now tracks all-time-high.

saw_081317_3 - Append latest action to market summary title.

saw_081317_2 - Make sure we don't try to follow ourselves!

saw_081317_1 - Fix pending list deletion index.

saw_081217_1 - Only follow users we're not already following. Add rudimentary bandwidth check.

saw_081117_2 - Use code that is compatible with older versions of matplotlib (e.g. 1.4.3).

saw_081117_1 - Market mode now has graphing!

saw_081017_4 - Use two decimal points for change percentage.

saw_081017_3 - Market mode now compares against previous summary.

saw_081017_2 - Update README.

saw_081017_1 - Add market mode.

saw_080917_1 - Fix daily wait check.

saw_080817_8 - Log daily vote number when upvoting.

saw_080817_7 - Remove another "post has vanished" warning.

saw_080817_6 - Timely now refers to Github URL and uses author-rooted permlink.

saw_080817_5 - Set a minimum curation payout of $0.50.

saw_080817_4 - Fix error in curate mode introduced in saw_080817_1.

saw_080817_3 - Exclude posts with 'boiled' tag.

saw_080817_2 - Allow other tags to be present in original timely posts.

saw_080817_1 - Reduce unnecessary output. Reset curation payout threshold after 20 hours even if
               not all votes have been made.

saw_080717_1 - Log the proper post when upvoting. Add curation section to README file.

saw_080517_1 - Fix thy self.

saw_080417_4 - Supply voter account id to post.upvote().

saw_080417_3 - More fixes for amount. Better exception handling.

saw_080417_2 - Fix for proper usage of Amount class.

saw_080417_1 - Add curate mode.

saw_073117_3 - Do not delete from dictionary while iterating.

saw_073117_2 - Use commit.post() instead of the buggy post.edit().

saw_073117_1 - Add timely mode.

saw_072717_2 - Make sure command script gets installed during `make install`.

saw_072717_1 - Boiler is functional!

saw_072617_3 - Add simple startup script and "run" Makefile target.

saw_072617_2 - Simple README.

saw_072617_1 - First commit for boiler.



ag/boiler/market.py
#!/usr/bin/env python3
# Copyright (C) 2017 Alpha Griffin
# @%@~LICENSE~@%@

from datetime import datetime, timedelta
from decimal import Decimal
from time import sleep
from tzlocal import get_localzone
from os import path
from urllib.parse import urlencode
from urllib.request import urlopen, Request
import base64
import json
import io

import matplotlib.pyplot as plt
from matplotlib.lines import Line2D, TICKLEFT, TICKRIGHT
from matplotlib.dates import DateFormatter
from steem import Steem
from steem.commit import Commit
from steembase.exceptions import RPCError

import ag.logging as log

from ag.boiler.__version__ import __version__
from ag.boiler.poloniex import Poloniex

from ag.boiler.config import account, mask, dir
log.debug("Account credentials loaded", id=account.id, key=mask(account.key))


class Market():

    testing = False

    def __init__(self, commit, api, pair, max_tries=60):
        self.commit = commit
        self.api = api
        self.max_tries = max_tries

        pair = pair.split('-')
        if len(pair) != 2:
            raise ValueError("Currency pair must be of the form XXX-YYY")

        self.symbol = pair[0].upper()
        self.against = pair[1].upper()

    def summarize(self, title, tags):
        log.info("Summarizing market...", symbol=self.symbol, against=self.against)

        if self.testing:
            log.info("TESTING MODE ENABLED")

        ticker = self.api.ticker()
        try:
            ticker = ticker[self.against + '_' + self.symbol]
        except KeyError as e:
            log.error("Currency pair not found in ticker data", symbol=self.symbol, against=self.against, exception=e)
            raise ValueError("Currency pair not found in ticker data")

        tz = get_localzone()
        now = datetime.now(tz)
        nowstr = now.strftime('%Y-%m-%d %H:%M:%S %Z')
        log.debug("got ticker data", now=nowstr, ticker=ticker)

        last = Decimal(ticker['last'])
        if self.against == 'USDT' or self.against == 'USD':
            symbol = '$'
            quant = Decimal('0.00')
        else:
            symbol = ''
            quant = Decimal('0.00000000')
        laststr = symbol + str(last.quantize(quant))
        log.debug("last trade", value=laststr)

        ath = None
        newath = False

        nowfile = path.join(dir, 'market.'+self.symbol+'-'+self.against+'.time')
        lastfile = path.join(dir, 'market.'+self.symbol+'-'+self.against+'.last')
        img_url = None

        if path.exists(nowfile) and path.exists(lastfile):
            prev = True

            with open(nowfile, 'r') as infile:
                prev_now = datetime.fromtimestamp(int(infile.readline().strip()), tz=tz)

            with open(lastfile, 'r') as infile:
                prev_last = Decimal(infile.readline().strip())

            prev_permlink = self.make_permlink(prev_now)
            prev_nowstr =   prev_now.strftime('%Y-%m-%d %H:%M:%S %Z')

            change_price = last - prev_last
            if change_price < Decimal('0'):
                change_pricestr = symbol + str(change_price.copy_negate().quantize(quant))
            else:
                change_pricestr = symbol + str(change_price.quantize(quant))

            change_pct = (Decimal('100') * change_price / prev_last).quantize(Decimal('0.00'))
            if change_pct < Decimal('0'):
                change_pctstr = str(change_pct.copy_negate()) + '%'
            else:
                change_pctstr = str(change_pct) + '%'

            highest = last
            lowest = last

            fig = plt.figure(figsize=(10,7), facecolor='k')
            ax = fig.add_subplot(1,1,1)
            rect = ax.patch
            rect.set_facecolor('k')
            img_title = self.symbol + '-' + self.against + ' at ' + nowstr
            plt.title(img_title)
            ax.xaxis_date()
            plt.xticks(rotation=25)
            ax.xaxis.set_major_formatter(DateFormatter('%Y-%m-%d %H:%M'))

            # first graph 30-minute candlesticks
            log.info("Plotting 30-minute candlesticks...")

            data = self.api.chartData(
                    pair=self.against+'_'+self.symbol,
                    start=int(prev_now.strftime("%s"))+1,
                    period=1800
                    )

            if len(data) < 0:
                raise ValueError("No data returned")
            elif len(data) == 1:
                try:
                    error = data['error']
                    log.error("Received error from API", error=error)
                    raise ValueError("Received error from API: {}".format(error))

                except KeyError:
                    if int(data[0]['date']) == 0:
                        raise ValueError("Too soon! You must wait at least 30 minutes between summaries for candlesticks.")

            for row in data:
                high = Decimal(row['high'])
                if high > highest:
                    highest = high

                low = Decimal(row['low'])
                if low < lowest:
                    lowest = low

                time = datetime.fromtimestamp(int(row['date']))
                popen = Decimal(row['open'])
                close = Decimal(row['close'])

                if close >= popen:
                    color = 'g'
                else:
                    color = 'r'

                vline = Line2D(xdata=(time, time), ydata=(low, high),
                        linewidth=1.5, color=color, antialiased=False)
                oline = Line2D(xdata=(time, time), ydata=(popen, popen),
                        linewidth=1, color=color, antialiased=False, marker=TICKLEFT, markersize=7)
                cline = Line2D(xdata=(time, time), ydata=(close, close),
                        linewidth=1, color=color, antialiased=False, marker=TICKRIGHT, markersize=7)

                ax.add_line(vline)
                ax.add_line(oline)
                ax.add_line(cline)

            # then graph 5-minute lines
            log.info("Plotting 5-minute lines...")

            data = self.api.chartData(
                    pair=self.against+'_'+self.symbol,
                    start=int(prev_now.strftime("%s"))+1,
                    period=300
                    )

            if len(data) < 0:
                raise ValueError("No data returned")
            elif len(data) == 1:
                try:
                    error = data['error']
                    log.error("Received error from API", error=error)
                    raise ValueError("Received error from API: {}".format(error))

                except KeyError:
                    if int(data[0]['date']) == 0:
                        raise ValueError("Too soon! You must wait at least 5 minutes between summaries.")

            begin = None

            for row in data:
                high = Decimal(row['high'])
                if high > highest:
                    highest = high

                low = Decimal(row['low'])
                if low < lowest:
                    lowest = low

                time = int(row['date'])
                popen = Decimal(row['open'])
                close = Decimal(row['close'])

                if begin is None:
                    begin = popen

                line = Line2D(xdata=(datetime.fromtimestamp(time), datetime.fromtimestamp(time + 300)), ydata=(begin, close),
                        linewidth=0.7, color='#FFFF00', antialiased=True)

                ax.add_line(line)
                begin = close

            higheststr = symbol + str(highest.quantize(quant))
            loweststr = symbol + str(lowest.quantize(quant))

            athfile = path.join(dir, 'market.'+self.symbol+'-'+self.against+'.ath')
            if path.exists(athfile):
                with open(athfile, 'r') as infile:
                    ath = Decimal(infile.readline().strip())

                if highest > ath:
                    ath = highest
                    newath = True

                    if not testing:
                        with open(athfile, 'w') as out:
                            out.write(str(ath))

            ax.xaxis.grid(True, color='#555555', linestyle='dotted')
            ax.yaxis.grid(True, color='#555555', linestyle='solid')
            plt.tight_layout()
            ax.autoscale_view()

            # save image to file or memory buffer
            if self.testing:
                imgfile = '/tmp/' + self.symbol + '-' + self.against + '.png'
                fig.savefig(imgfile)
                log.info("Market graph PNG saved", file=imgfile)
            else:
                img = io.BytesIO()
                fig.savefig(img, format='png')
                img.seek(0)

            plt.close(fig)

            # now upload result to imgur
            if not self.testing:
                log.info("Uploading plot to imgur...")

                img_b64 = base64.standard_b64encode(img.read())
                client = 'bbe2ecf93d88915'
                headers = {'Authorization': 'Client-ID ' + client}
                imgur_data = {'image': img_b64, 'title': img_title}
                req = Request(url='https://api.imgur.com/3/upload.json', data=urlencode(imgur_data).encode('ASCII'), headers=headers)
                resp = urlopen(req).read()
                resp = json.loads(resp)
                log.debug("Got response from imgur", resp=resp)

                if resp['success'] == True:
                    img_url = resp['data']['link']
                    log.info("Image uploaded successfully", url=img_url)

                else:
                    log.error("Non-successful response from imgur", resp=resp)
                    raise ValueError("Non-successful response from imgur")

        else:
            prev = False

        body = "Market Summary for " + self.symbol
        body += "\n=="
        body += "\n* All prices in *" + self.against + "*"
        body += "\n---"
        body += "\n"
        if prev:
            if change_pct > Decimal('0'):
                body += "\nUp " + change_pctstr
                title += ": Up " + change_pctstr
            elif change_pct < Decimal('0'):
                body += "\nDown " + change_pctstr
                title += ": Down " + change_pctstr
            else:
                body += "\nFlat"
                title += ": Flat"
            if newath:
                body += " (New All Time High Achieved)"
                title += " -- New All Time High!"
            body += "\n-"
            body += "\n" + self.symbol + " **"
            if change_price > Decimal('0'):
                body += "gained " + change_pricestr
            elif change_price < Decimal('0'):
                body += "lost " + change_pricestr
            else:
                body += "had no change"
            body += "** since the [last market summary]"
            body += "(https://steemit.com/@" + account.id + "/" + prev_permlink + ")"
            if change_pct > Decimal('0'):
                body += ", a change of **" + change_pctstr + "**"
            elif change_pct < Decimal('0'):
                body += ", a change of **-" + change_pctstr + "**"
            body += "."
        else:
            body += "\n*This is the first market summary, so no previous comparison data is available.*"
        body += "\n"
        body += "\n* Last trade: *" + laststr + "*"
        if prev:
            body += "\n* Highest trade: *" + higheststr + "*"
            if newath:
                body += " (new all time high)"
            body += "\n* Lowest trade: *" + loweststr + "*"
            if img_url is not None:
                body += "\n"
                body += "\n[![market activity plot](" + img_url + ")](" + img_url + ")"
        body += "\n"
        body += "\n---"
        body += "\n"
        body += "\n* Snapshot taken at *" + nowstr + "*"
        if prev:
            body += "\n* Previous snapshot: *[" + prev_nowstr + "]"
            body += "(https://steemit.com/@" + account.id + "/" + prev_permlink + ")*"
        body += "\n* Quote data from [Poloniex](http://poloniex.com)"
        body += "\n"
        body += "\n<center>Happy trading... stay tuned for the next summary!</center>"
        body += "\n"
        body += "\n---"
        body += "\n<center>*This market summary produced automatically by:"
        body += "\n[![Alpha Griffin logo](http://alphagriffin.com/usr/include/ag/favicon/favicon-128.png)"
        body += "\nAlpha Griffin Boiler bot](https://github.com/AlphaGriffin/boiler)"
        body += "\nv" + __version__ + "*</center>"

        if self.testing:
            print(body)

        permlink = self.make_permlink(now)
        tries = 0
        post = None

        while tries < self.max_tries:
            try:
                log.info("Posting summary...", permlink=permlink, title=title, last=laststr, tags=tags)

                if self.testing:
                    log.warn("Not actually going to post (testing mode)")
                    break

                post = self.commit.post(
                        permlink = permlink,
                        title = title,
                        author = account.id,
                        body = body,
                        tags = tags,
                        self_vote = True
                        )

                break

            except RPCError as e:
                log.warn("Got RPC error while posting, trying again in 1 minute...", exception=e)
                tries += 1
                sleep(60)

        if post is not None:
            log.info("Summary posted successfully", post=post)

            with open(nowfile, 'w') as out:
                out.write(now.strftime("%s"))

            with open(lastfile, 'w') as out:
                out.write(str(last))

            return True

        else:
            if not self.testing:
                log.error("Failed to post summary")

            return False

    def make_permlink(self, when):
        permlink = 'market-summary-' + self.symbol + '-' + self.against + '-' + when.strftime('%Y-%m-%d-%H-%M-%S-%Z')
        permlink = permlink.lower()  # STEEM permlinks must be all lowercase
        return permlink


def all_time_high(pair, price):
    pair = pair.split('-')
    if len(pair) != 2:
        raise ValueError("Currency pair must be of the form XXX-YYY")

    symbol = pair[0].upper()
    against = pair[1].upper()

    price = Decimal(price)
    file = path.join(dir, 'market.'+symbol+'-'+against+'.ath')

    log.info("Saving new all-time-high...", symbol=symbol, against=against, price=price, file=file)

    with open(file, 'w') as out:
        out.write(str(price))

    print("All-time-high saved successfully.")


def run(args):
    log.info("Market summary mode activated", args=args)

    if args is None or len(args) < 3:
        raise ValueError("You must specify a currency pair, title, and one or more tags")

    pair = args[0]
    title = args[1]
    tags = args[2:]

    log.debug("initializing...")
    steem = Steem(keys=[account.key])
    commit = Commit(steem)
    api = Poloniex()
    market = Market(commit, api, pair)
    log.debug("ready", steem=steem, commit=commit, api=api, market=market)

    market.summarize(title, tags)


if __name__ == '__main__':
    from contextlib import suppress
    from sys import argv

    with suppress(KeyboardInterrupt):
        run(argv[1:])

